pkg/operator/image.go

```diff
@@ -0,0 +1,130 @@
+	Image            *string
+	DefaultBaseImage string
+	BaseImage        string
+	DefaultVersion   string
+	Version          string
+	Tag              string
+	SHA              string
+}
+
+// ContainerImageURL determines the appropriate container image URL based
+// on the data coming from default values, operator CLI args, and user
+// configured fields in the CRD.
+//
+// If the `Image` field is not empty, this takes priority over other
+// settings.  If the given `Image` URL does not contain a version,
+// then the configured version or the default version is appended to
+// the image URL.
+//
+// If no `Image` is specified, an image URL is created by combining the
+// appropriate `BaseImage` with the `Version`, `Tag`, and `SHA`, fields.
+//
+// Returns the image URL of the component.
+func ContainerImageURL(c *ImageConfig) string {
+	version := c.DefaultVersion
+	if c.Version != "" {
+		version = c.Version
+	}
+
+	if c.Image != nil && *c.Image != "" {
+		return ApplyImageVersion(*c.Image, version)
+	}
+
+	// Version is used by default.
+	// If the tag is specified, we use the tag to identify the container image.
+	// If the sha is specified, we use the sha to identify the container image,
+	// as it has even stronger immutable guarantees to identify the image.
+	baseImage := c.DefaultBaseImage
+	if c.BaseImage != "" {
+		baseImage = c.BaseImage
+	}
+	image := fmt.Sprintf("%s:%s", baseImage, version)
+	if c.Tag != "" {
+		image = fmt.Sprintf("%s:%s", baseImage, c.Tag)
+	}
+	if c.SHA != "" {
+		image = fmt.Sprintf("%s@sha256:%s", baseImage, c.SHA)
+	}
+	return image
+}
+
+// ChooseVersion returns the parsed semver.Version from one of the three given
+// string parameters.  If `version` includes a non-empty string, it is returned.
+// Otherwise, try to get a valid version from the `imageURL`, and lastly, return
+// default version if the others are not successful.
+func ChooseVersion(version, imageURL, defaultVersion string) (semver.Version, error) {
+	if version != "" {
+		return semver.ParseTolerant(version)
+	}
+	if imageVersion := GetImageVersion(imageURL); imageVersion != "" {
+		return semver.ParseTolerant(imageVersion)
+	}
+	return semver.ParseTolerant(defaultVersion)
+}
+
+const (
+	imageHostRegex    = `[\w-.]+(:[\d]+)?/`
+	imageNameRegex    = `([\w-.]+/)?[\w-.]+`
+	imageVersionRegex = `([@][\w]+)?:[\w-.]+`
+)
+
+var (
+	imageRegexStr = fmt.Sprintf(`%s%s%s`, imageHostRegex, imageNameRegex, imageVersionRegex)
+	imageRegex    = regexp.MustCompile(`^(` + imageHostRegex + `)?` + imageNameRegex + `(` + imageVersionRegex + `)?$`)
+)
+
+// GetImageVersion gets the version/tag portion of a container
+// image URL.  If there is no version specified, or if the url
+// specifies a digest instead of a version or tag, an empty string
+// is returned.
+func GetImageVersion(image string) string {
+	match := imageRegex.FindStringSubmatch(image)
+	if len(match) < 6 {
+		return ""
+	}
+	fmt.Println("image: " + image + ", matched: " + match[4])
```

Comment:
I assume this wasn't meant to be left here

Reply:
Yes, accidental debugging statement, thanks!